File: objc/Project12/Project12 WatchKit Extension/ComplicationController.m
//
//  ComplicationController.m
//  Project12 WatchKit Extension
//
//  Created by Jinwoo Kim on 2/2/21.
//

#import "ComplicationController.h"

/// 6번
@implementation ComplicationController

- (void)getSupportedTimeTravelDirectionsForComplication:(CLKComplication *)complication withHandler:(void (^)(CLKComplicationTimeTravelDirections))handler {
    handler(0);
}

- (void)getComplicationDescriptorsWithHandler:(void (^)(NSArray<CLKComplicationDescriptor *> * _Nonnull))handler {
    NSArray<CLKComplicationDescriptor *> *descriptors = @[
        [[CLKComplicationDescriptor alloc] initWithIdentifier:@"complication"
                                                  displayName:@"Project12"
                                            supportedFamilies:@[[NSNumber numberWithInt:CLKComplicationFamilyModularSmall]]]
    ];
    
    handler(descriptors);
}

- (void)getCurrentTimelineEntryForComplication:(CLKComplication *)complication withHandler:(void (^)(CLKComplicationTimelineEntry * _Nullable))handler {
    NSString *currentText = [NSUserDefaults.standardUserDefaults valueForKey:@"complication_number"];
    if (currentText == nil) currentText = @"❤️";
    CLKComplicationTemplateModularSmallSimpleText *template = [CLKComplicationTemplateModularSmallSimpleText templateWithTextProvider:[CLKTextProvider textProviderWithFormat:@"%@", currentText]];
    CLKComplicationTimelineEntry *entry = [CLKComplicationTimelineEntry entryWithDate:[NSDate new] complicationTemplate:template];
    handler(entry);
}

- (void)getLocalizableSampleTemplateForComplication:(CLKComplication *)complication withHandler:(void (^)(CLKComplicationTemplate * _Nullable))handler {
    // This method will be called once per supported complication, and the results will be cached
    CLKComplicationTemplateModularSmallSimpleText *template = [CLKComplicationTemplateModularSmallSimpleText templateWithTextProvider:[CLKTextProvider textProviderWithFormat:@"💔"]];
    handler(template);
}

@end

File: swift/Project12/Project12 WatchKit Extension/ComplicationController.swift
//
//  ComplicationController.swift
//  Project12 WatchKit Extension
//
//  Created by Jinwoo Kim on 2/2/21.
//

import ClockKit

/// 6번
class ComplicationController: NSObject, CLKComplicationDataSource {
    func getSupportedTimeTravelDirections(for complication: CLKComplication, withHandler handler: @escaping (CLKComplicationTimeTravelDirections) -> Void) {
        handler([])
    }
    
    func getComplicationDescriptors(handler: @escaping ([CLKComplicationDescriptor]) -> Void) {
        let descriptors = [
            CLKComplicationDescriptor(identifier: "complication", displayName: "Project12", supportedFamilies: [.modularSmall])
        ]
        handler(descriptors)
    }
    
    func getCurrentTimelineEntry(for complication: CLKComplication, withHandler handler: @escaping (CLKComplicationTimelineEntry?) -> Void) {
        let currentText = UserDefaults.standard.string(forKey: "complication_number") ?? "❤️"
        let template = CLKComplicationTemplateModularSmallSimpleText(textProvider: CLKTextProvider(format: currentText))
        let entry = CLKComplicationTimelineEntry(date: Date(), complicationTemplate: template)
        handler(entry)
    }
    
    func getLocalizableSampleTemplate(for complication: CLKComplication, withHandler handler: @escaping (CLKComplicationTemplate?) -> Void) {
        // This method will be called once per supported complication, and the results will be cached
        let template = CLKComplicationTemplateModularSmallSimpleText(textProvider: CLKTextProvider(format: "💔"))
        handler(template)
    }
}
